src/parse.jl

```diff
@@ -101,3 +102,523 @@ function ZonedDateTime(str::AbstractString, format::AbstractString; locale::Abst
 end
 
 Dates.default_format(::Type{ZonedDateTime}) = ISOZonedDateTimeFormat
+
+
+"""
+    _parsesub_tzabbr(str, [i, len]) -> Union{Tuple{AbstractString, Integer}, Exception}
+
+Parses a section of a string as a time zone abbreviation as defined in the tzset(3) man
+page. An abbreviation must be at least three or more alphabetic characters. When enclosed by
+the less-than (<) and greater-than (>) signs the character set is expanded to include the
+plus (+) sign, the minus (-) sign, and digits.
+
+# Examples
+```jldoctest
+julia> _parsesub_tzabbr("ABC+1")
+("ABC", 4)
+
+julia> _parsesub_tzabbr("<ABC+1>+2")
+("ABC+1", 6)
+```
+"""
+function _parsesub_tzabbr(
+    str::AbstractString,
+    i::Integer=firstindex(str),
+    len::Integer=lastindex(str),
+)
+    state = :started
+    name_start = i
+    name_end = prevind(str, i)
+
+    @inbounds while i <= len
+        c, ii = iterate(str, i)::Tuple{Char, Int}
+        isascii(c) || break  # Restricts abbreviation support to ASCII characters
+
+        if state == :simple && isletter(c)
+            name_end = i
+        elseif state == :expanded && (isletter(c) || isdigit(c) || c in ('+', '-'))
+            name_end = i
+        elseif state == :started && c == '<'
+            name_start = ii
+            name_end = i
+            state = :expanded
+        elseif state == :started && isletter(c)
+            name_end = i
+            state = :simple
+        elseif state == :expanded && c == '>'
+            i = ii
+            state = :closed
+            break
+        else
+            break
+        end
+        i = ii
+    end
+
+    if state == :expanded
+        return ParseNextError("Expected expanded time zone abbreviation end with the greater-than sign (>)", str, prevind(str, name_start), i)
+    elseif state != :closed && name_start > name_end
+        return ParseNextError("Time zone abbreviation must start with a letter or the less-than (<) character", str, i)
+    elseif length(str, name_start, name_end) < 3
+        char_type = if state == :simple
+            "alphabetic characters"
+        else
+            "characters which are either alphanumeric, the plus sign (+), or the minus sign (-)"
+        end
+
+        return ParseNextError("Time zone abbreviation must be at least three $char_type", str, name_start, name_end)
+    else
+        name = SubString(str, name_start, name_end)
+        return name, i
+    end
+end
+
+
+"""
+    _parsesub_offset(str, [i, len]) -> Union{Tuple{Integer, Integer}, Exception}
```

Comment:
We might want to start using ResultTypes in this package for type stability? 

https://github.com/iamed2/ResultTypes.jl

Reply:
I was avoiding using other packages here mainly as I want to use this pattern in the Dates stdlib. However I should do some benchmarking using ResultTypes

Reply:
I'll note that with newer versions of Julia there isn't much a performance impact for using `Union`s. However there may be performance improvements still by using ResultTypes. I have the beginnings of some benchmarks using PkgBenchmarks which once established it would be good to try out ResultTypes to see if I can get additional speedups. With all of that said I think I'll just continue to sit on this PR if I continue to wait for the rest of my experimental work.

I'll leave this unresolved for now and I'll be sure to reference the ResultTypes benchmarks here.